Source/Core/Core/IOS/USB/USBV5.cpp

```diff
@@ -19,39 +21,257 @@ namespace HLE
+    return GetDefaultReply(IPC_EINVAL);
+
+  std::lock_guard<std::mutex> lock{m_usbv5_devices_mutex};
+  USBV5Device* device = GetUSBV5Device(request.buffer_in);
+  if (!device)
+    return GetDefaultReply(IPC_EINVAL);
+  return handler(*device);
+}
+
+void USBV5ResourceManager::OnDeviceChange(const ChangeEvent event,
+                                          std::shared_ptr<USB::Device> device)
+{
+  std::lock_guard<std::mutex> lock{m_usbv5_devices_mutex};
+  const u64 host_device_id = device->GetId();
+  if (event == ChangeEvent::Inserted)
+  {
+    for (const auto& interface : device->GetInterfaces(0))
+    {
+      if (interface.bAlternateSetting != 0)
+        continue;
+
+      auto it = std::find_if(m_usbv5_devices.rbegin(), m_usbv5_devices.rend(),
+                             [](const USBV5Device& entry) { return !entry.in_use; });
+      if (it == m_usbv5_devices.rend())
+        return;
+
+      it->in_use = true;
+      it->interface_number = interface.bInterfaceNumber;
+      it->number = m_current_device_number;
+      it->host_id = host_device_id;
+    }
+  }
+  else if (event == ChangeEvent::Removed)
+  {
+    for (USBV5Device& entry : m_usbv5_devices)
+    {
+      if (entry.host_id == host_device_id)
+        entry.in_use = false;
+    }
+  }
+}
+
+void USBV5ResourceManager::OnDeviceChangeEnd()
+{
+  std::lock_guard<std::mutex> lk{m_devicechange_hook_address_mutex};
+  TriggerDeviceChangeReply();
+  ++m_current_device_number;
+}
+
+// Must be called with m_devicechange_hook_address_mutex locked
+void USBV5ResourceManager::TriggerDeviceChangeReply()
+{
+  if (!m_devicechange_hook_request)
+    return;
+
+  std::lock_guard<std::mutex> lock{m_usbv5_devices_mutex};
+  u8 num_devices = 0;
+  for (auto it = m_usbv5_devices.crbegin(); it != m_usbv5_devices.crend(); ++it)
+  {
+    const USBV5Device& usbv5_device = *it;
+    if (!usbv5_device.in_use)
+      continue;
+
+    const auto device = GetDeviceById(usbv5_device.host_id);
+    if (!device)
+      continue;
+
+    DeviceEntry entry;
+    if (HasInterfaceNumberInIDs())
+    {
+      entry.id.reserved = usbv5_device.interface_number;
+    }
+    else
+    {
+      // The actual value is static_cast<u8>(hook_internal_ipc_request >> 8).
+      // Since we don't actually emulate the IOS kernel and internal IPC,
+      // just pretend the value is 0xe7 (most common value according to hwtests).
+      entry.id.reserved = 0xe7;
+    }
+    entry.id.index = std::distance(m_usbv5_devices.cbegin(), it.base()) - 1;
+    entry.id.number = Common::swap16(usbv5_device.number);
+    entry.vid = Common::swap16(device->GetVid());
+    entry.pid = Common::swap16(device->GetPid());
+    entry.number = Common::swap16(usbv5_device.number);
+    entry.interface_number = usbv5_device.interface_number;
+    entry.num_altsettings = device->GetNumberOfAltSettings(entry.interface_number);
+
+    Memory::CopyToEmu(m_devicechange_hook_request->buffer_out + sizeof(entry) * num_devices++,
```

Comment:
`num_devices++` can come after the `CopyToEmu` call, no? Totally opinion-based, but it seems more readable in that case.